File: server/src/entities/Post.ts
import { Field, Int, ObjectType } from 'type-graphql';
import {
  BaseEntity,
  Column,
  CreateDateColumn,
  Entity,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from 'typeorm';
import { Comment } from './Comment';
import { Community } from './Community';
import { Upvote } from './Upvote';
import { User } from './User';

@ObjectType()
@Entity()
export class Post extends BaseEntity {
  @Field()
  @PrimaryGeneratedColumn()
  id!: number;

  @Field()
  @Column()
  title!: string;

  @Field()
  @Column({ default: '' })
  text: string;

  @Field()
  @Column({ default: '' })
  image: string;

  @Field()
  @Column({ default: '' })
  link: string;

  @Field()
  @Column({ type: 'int', default: 0 })
  points!: number;

  @Field(() => Int, { nullable: true })
  voteStatus: number | null;

  @Field()
  @Column()
  creatorId: number;

  @Field(() => User)
  @ManyToOne(() => User, (user) => user.posts, { onDelete: 'CASCADE' })
  creator: User;

  @Field()
  @Column()
  communityId: number;

  @ManyToOne(() => Community, (community) => community.posts)
  community: Community;

  @Field(() => [Comment])
  @OneToMany(() => Comment, (comment) => comment.post)
  comments: Comment[];

  @OneToMany(() => Upvote, (upvote) => upvote.post)
  upvotes: Upvote[];

  @Field(() => String)
  @CreateDateColumn()
  createdAt: Date;

  @Field(() => String)
  @UpdateDateColumn()
  updatedAt: Date;
}

File: server/models/post.go
package models

import (
	"time"
)

type Post struct {
	ID          int `gorm:"primaryKey"`
	CreatedAt   time.Time
	UpdatedAt   time.Time
	Title       string    `json:"title"`
	Text        string    `json:"text"`
	Points      int       `json:"points" gorm:"default: 0"`
	CreatorID   int       `json:"creatorId"`
	Creator     User      `json:"creator"`
	CommunityID int       `json:"communityId"`
	Community   Community `json:"community"`
	Upvotes     []Upvote  `json:"upvotes"`
}

type PaginatedPosts struct {
	Posts   []Post
	HasMore bool
}

type PostInput struct {
	Title string `json:"title"`
	Text  string `json:"text"`
}
